### PR TITLE
Add term frequency functions to score script query

### LINKED ISSUE #4858: [DOC] Document new doc/term frequency functions in Painless score scripts
**What do you want to do?**
 
- [ ] Request a change to existing documentation
- [x] Add new documentation
- [ ] Report a technical problem with the documentation
- [ ] Other 

**Tell us about your request.** Provide a summary of the request and all versions that are affected.
Starting in version 2.10, OpenSearch will support four new frequency functions to be used in Painless score scripts. While this is likely to be an expert feature only, we should document their existence and what they do.

Note that these functions should show equivalent behavior to the similarly named functions in Apache Solr, since the underlying Lucene implementation is the same: https://solr.apache.org/guide/solr/latest/query-guide/function-queries.html

**What other resources are available?** Provide links to related issues, POCs, steps for testing, etc.

* https://github.com/opensearch-project/OpenSearch/pull/9081
* https://github.com/opensearch-project/OpenSearch/issues/7558
* https://github.com/opensearch-project/OpenSearch/issues/8702

### PR DESCRIPTION
**Pull request by kolchfa-aws**
Fixes #4858 

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _query-dsl/specialized/script-score.md [hunk 1]
```diff
@@ -328,5 +328,52 @@ GET blogs/_search
 ```
 {% include copy-curl.html %}
 
+### Term frequency functions
+
+Term frequency functions expose term-level statistics in a script score context. You can use these statistics to implement custom information retrieval and ranking algorithms like query-time multiplicative or additive score boosting by popularity. To apply a term frequency function, call one of the following Painless methods:
```
Comment (noCharger) [inline _query-dsl/specialized/script-score.md:333]: There's `script context` https://github.com/noCharger/OpenSearch/blob/main/server/src/main/java/org/opensearch/script/ScriptContext.java and each script has their own context instance https://github.com/noCharger/OpenSearch/blob/main/server/src/main/java/org/opensearch/script/ScoreScript.java#L287. I would say expose them in score script `source`.

File: _query-dsl/specialized/script-score.md [hunk 2]
```diff
@@ -328,5 +328,52 @@ GET blogs/_search
...
+
+- `int termFreq(String <field-name>, String <term>)`: Retrieves the term frequency within a field for a specific term.
+- `float tf(String <field-name>, String <term>)`: Calculates the term frequency/inverse document frequency (TF/IDF) for a specific term within a field.
```
Comment (noCharger) [inline _query-dsl/specialized/script-score.md:336]: There's some setup for using `tf`, without them you will get errors like this:

```
{
  "error": {
    "root_cause": [
      {
        "type": "unsupported_operation_exception",
        "reason": "unsupported_operation_exception: requires a TFIDFSimilarity (such as ClassicSimilarity)"
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "test",
        "node": "sQkgvgQWRT2pfI4leTC8lg",
        "reason": {
          "type": "script_exception",
          "reason": "runtime error",
          "script_stack": [
            "org.opensearch.ExceptionsHelper.convertToOpenSearchException(ExceptionsHelper.java:86)",
            "org.opensearch.script.ScoreScriptUtils$TF.tf(ScoreScriptUtils.java:113)",
            "tf(params.field, params.term)",
            "                       ^---- HERE"
          ],
          "script": "tf(params.field, params.term)",
          "lang": "painless",
          "position": {
            "offset": 23,
            "start": 0,
            "end": 29
          },
          "caused_by": {
            "type": "exception",
            "reason": "java.lang.UnsupportedOperationException: requires a TFIDFSimilarity (such as ClassicSimilarity)",
            "caused_by": {
              "type": "unsupported_operation_exception",
              "reason": "unsupported_operation_exception: requires a TFIDFSimilarity (such as ClassicSimilarity)"
            }
          }
        }
      }
    ],
    "caused_by": {
      "type": "unsupported_operation_exception",
      "reason": "unsupported_operation_exception: requires a TFIDFSimilarity (such as ClassicSimilarity)"
    }
  },
  "status": 400
}
```

Here's some suggestions.

1. Setup the Similarity Model:
You need to set the similarity model of the field you're working with Ex: ClassicSimilarity, TFIDFSimilarity). Make sure the model is supported under current OpenSearch Version

To do this, you can update the mappings of your index:

```
PUT /test/_mapping
{
  "properties": {
    "<field_name>": {
      "type": "text",
      "similarity": "classic"
    }
  }
}
```

2. Reindex your data for existing indices

File: _query-dsl/specialized/script-score.md [hunk 3]
```diff
@@ -328,5 +328,52 @@ GET blogs/_search
...
+              }
+            }
+            return params.default_value;
+            
+          """,
+          "params": {
+              "fields": ["title", "description"],
+              "term": "ai",
+              "multiplier": 2,
+              "default_value": 1
```
Comment (noCharger) [inline _query-dsl/specialized/script-score.md:368]: nit picking: it's more consistent use the same doc schema for all examples in this page

File: _query-dsl/specialized/script-score.md (hunk 2 above)
Comment (kolchfa-aws) [reply _query-dsl/specialized/script-score.md:336]: When I tried to set similarity to "classic", I got an error that suggests to use "BM25" as similarity: 

```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "The [classic] similarity may not be used anymore. Please use the [BM25] similarity or build a custom [scripted] similarity instead."
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "The [classic] similarity may not be used anymore. Please use the [BM25] similarity or build a custom [scripted] similarity instead."
  },
  "status": 400
}
```

Comment (noCharger) [reply _query-dsl/specialized/script-score.md:336]: Good catch, would you create an issue in OpenSearch repo or I can do it too.

**Review by noCharger — APPROVED** [review]

File: _query-dsl/specialized/script-score.md (hunk 2 above)
Comment (kolchfa-aws) [reply _query-dsl/specialized/script-score.md:336]: I can create an issue

Comment (kolchfa-aws) [reply _query-dsl/specialized/script-score.md:336]: Done. The OpenSearch issue is https://github.com/opensearch-project/OpenSearch/issues/9958

File: _query-dsl/specialized/script-score.md [hunk 4]
```diff
@@ -328,5 +328,73 @@ GET blogs/_search
...
+Term frequency functions expose term-level statistics in the score script source. You can use these statistics to implement custom information retrieval and ranking algorithms like query-time multiplicative or additive score boosting by popularity. To apply a term frequency function, call one of the following Painless methods:
+
+- `int termFreq(String <field-name>, String <term>)`: Retrieves the term frequency within a field for a specific term.
+- `float tf(String <field-name>, String <term>)`: Calculates the term frequency/inverse document frequency (TF/IDF) for a specific term within a field.
+- `long totalTermFreq(String <field-name>, String <term>)`: Retrieves the total term frequency within a field for a specific term.
+- `long sumTotalTermFreq(String <field-name>)`: Retrieves the sum of total term frequencies within a field.
+
+#### Using `tf`
+
+Before using `tf`, you must set the similarity model on the field for which you want to retrieve term frequency:
```
Comment (cwillum) [inline _query-dsl/specialized/script-score.md:342]: Did I keep the meaning here?
```suggestion
Before using `tf`, you must set the similarity model for the field whose term frequency you want to retrieve:
```

**Review by cwillum — APPROVED** [review]
Looks good. Just one comment from me.

File: _query-dsl/specialized/script-score.md [hunk 5]
```diff
@@ -328,5 +328,73 @@ GET blogs/_search
 ```
 {% include copy-curl.html %}
 
+### Term frequency functions
+
+Term frequency functions expose term-level statistics in the score script source. You can use these statistics to implement custom information retrieval and ranking algorithms like query-time multiplicative or additive score boosting by popularity. To apply a term frequency function, call one of the following Painless methods:
```
Comment (natebower) [inline _query-dsl/specialized/script-score.md:333]:
```suggestion
Term frequency functions expose term-level statistics in the score script source. You can use these statistics to implement custom information retrieval and ranking algorithms, like query-time multiplicative or additive score boosting by popularity. To apply a term frequency function, call one of the following Painless methods:
```

File: _query-dsl/specialized/script-score.md [hunk 6]
```diff
@@ -328,5 +328,73 @@ GET blogs/_search
...
+    }
+  }
+  ```
+  {% include copy-curl.html %}
+
+1. Reindex your data for the existing indexes.
+
+#### Example
+
+The following query calculates the score as the total term frequency for each field in the `fields` list, multiplied by the `multiplier` value:
```
Comment (natebower) [inline _query-dsl/specialized/script-score.md:363]:
```suggestion
The following query calculates the score as the total term frequency for each field in the `fields` list multiplied by the `multiplier` value:
```

**Review by natebower — COMMENTED** [review]
@kolchfa-aws Just a couple of commas. Otherwise, LGTM!

**Comment by noCharger** [comment]
let's remove the docs for tf(). https://github.com/opensearch-project/OpenSearch/pull/9995

**Comment by noCharger** [comment]
Hi team, anything can I do to get this PR merged?

**Comment by kolchfa-aws** [comment]
@noCharger The PR is in the "Done but ready to merge" status. This means we'll merge it by the release date. So you can mark it done on your side.

**Pull request merged**